### PR TITLE
Reduce operator

### PR DESCRIPTION
**Pull request by CorentinBT**

### CODE REVIEW

File: src/rpp/rpp/operators/reduce.hpp [hunk 1]
```diff
@@ -55,10 +55,11 @@ struct reduce_t : public operators::details::operator_observable_strategy<reduce
     using ResultValue = Seed;
 };
 
-template<rpp::constraint::decayed_type Seed, rpp::constraint::observer TObserver, rpp::constraint::decayed_type Accumulator>
+template<rpp::constraint::observer TObserver, rpp::constraint::decayed_type Accumulator>
 struct reduce_no_seed_observer_strategy
 {
     using DisposableStrategyToUseWithThis = rpp::details::none_disposable_strategy;
+    using Seed = std::invoke_result_t<Accumulator, utils::convertible_to_any, utils::convertible_to_any>;
```
Comment (CorentinBT) [inline src/rpp/rpp/operators/reduce.hpp]: @victimsnino What do you think? Is it to remove Seed symbol from strategy template

Comment (victimsnino) [reply src/rpp/rpp/operators/reduce.hpp]: At this point Seed is rpp::utils::extract_observer_type_t<TObserver>

Comment (victimsnino) [reply src/rpp/rpp/operators/reduce.hpp]: But you proposal wouldn't work for templated lambdas

Comment (victimsnino) [reply src/rpp/rpp/operators/reduce.hpp]: Same about `reduce_observer_strategy` and same about scan operator..

Comment (CorentinBT) [reply src/rpp/rpp/operators/reduce.hpp]: Indeed, let me revert

Comment (victimsnino) [reply src/rpp/rpp/operators/reduce.hpp]: Could you update to rpp::utils::extract_observer_type_t everywhere instead?

File: src/examples/rpp/doxygen/reduce.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @example reduce.cpp
+ **/
+
+int main()
+{
+    //! [reduce]
+    rpp::source::just(1, 2, 3)
+        | rpp::operators::reduce(10, std::plus<int>{})
```
Comment (victimsnino) [inline src/examples/rpp/doxygen/reduce.cpp]: can we have reduce example with different types? like std::string

File: src/tests/rpp/test_reduce.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+#include <snitch/snitch.hpp>
+
+#include <rpp/operators/reduce.hpp>
+#include <rpp/sources/from.hpp>
+
+#include "copy_count_tracker.hpp"
+#include "mock_observer.hpp"
+
+TEMPLATE_TEST_CASE("scan scans values and store state", "", rpp::memory_model::use_stack, rpp::memory_model::use_shared)
```
Comment (victimsnino) [inline src/tests/rpp/test_reduce.cpp]: reduce reduces? =)

File: src/rpp/rpp/operators/reduce.hpp (hunk 1 above)
Comment (CorentinBT) [reply src/rpp/rpp/operators/reduce.hpp]: done

File: src/examples/rpp/doxygen/reduce.cpp (hunk 2 above)
Comment (CorentinBT) [reply src/examples/rpp/doxygen/reduce.cpp]: done

File: src/tests/rpp/test_reduce.cpp (hunk 3 above)
Comment (CorentinBT) [reply src/tests/rpp/test_reduce.cpp]: done

File: src/examples/rpp/doxygen/reduce.cpp [hunk 4]
```diff
@@ -9,17 +12,17 @@
 int main()
 {
     //! [reduce]
-    rpp::source::just(1, 2, 3)
-        | rpp::operators::reduce(10, std::plus<int>{})
-        | rpp::operators::subscribe([](int v) { std::cout << v << std::endl; });
-    // Output: 16
+    rpp::source::just("b"s, "c"s)
+        | rpp::operators::reduce("a"s, std::plus<std::string>{})
```
Comment (victimsnino) [inline src/examples/rpp/doxygen/reduce.cpp]: No, my original idea was to make reduce working with different types. Like getting observalbe of ints and then reducing it to concatenated std::string =)

Comment (victimsnino) [reply src/examples/rpp/doxygen/reduce.cpp]: But NVM =)

**Review by victimsnino — DISMISSED** [review]

**Pull request merged**